### PR TITLE
[CIT-364] remove null character

### PR DESCRIPTION
**Pull request by PradeepKiruvale**
## Proposed changes

This PR captures the code changes for removing the null character if the string is terminated with a null character.

## Types of changes

What types of changes does your code introduce to `thin-edge.io`?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (general improvements like code refactorings that doesn't explicitly fix a bug or add any new functionality)
- [ ] Documentation Update (if none of the other choices apply)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Paste Link to the issue
<br/>

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://github.com/thin-edge/thin-edge.io/blob/main/CONTRIBUTOR-LICENSE-AGREEMENT.md) (in all commits with git commit -s)
- [ ] I ran `cargo fmt` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I used `cargo clippy` as mentioned in [CODING_GUIDELINES](https://github.com/thin-edge/thin-edge.io/blob/main/CODING_GUIDELINES.md)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Review by didier-wenzek — COMMENTED** [review]
Don't forget to remove the current fix that has been done in `dm_agent/collectd`.

File: common/mqtt_client/src/lib.rs [hunk 1]
```diff
@@ -490,6 +490,13 @@ impl Message {
     pub fn qos(self, qos: QoS) -> Self {
         Self { qos, ..self }
     }
+
+    fn trim_if_null_terminated(payload: &mut Vec<u8>) -> &mut Vec<u8> {
+        if !payload.is_empty() && payload[payload.len() - 1].eq(&0) {
```
Comment (mneumann) [inline common/mqtt_client/src/lib.rs:495]: 1. Don't make this a function belonging to `Message`. It has nothing to do with a message. it just drops the NUL of a byte vector.

2. The implementation does one copy too much. Each time, it creates a clone of the vector in `to_vec`, which is unneccesarry.

Use this function instead:

```rust
fn trim_null_terminated(mut vec: Vec<u8>) -> Vec<u8> {
    if let Some(0) = vec.as_slice().last() {
        vec.pop();
    }
    vec
}
```

Comment (mneumann) [reply common/mqtt_client/src/lib.rs:495]: And then:

```rust
   payload: trim_null_terminated(payload.into())
```

File: common/mqtt_client/src/lib.rs [hunk 2]
```diff
@@ -686,4 +693,55 @@ mod tests {
         assert!(TopicFilter::new("/a/#/b").is_err());
         assert!(TopicFilter::new("/a/#/+").is_err());
     }
+
+    #[test]
+    fn check_null_terminated_messages() {
+        let mut payload1 = vec![155, 156, 157, 158, 0];
+        assert_eq!(
+            Message::trim_if_null_terminated(&mut payload1).to_vec(),
```
Comment (mneumann) [inline common/mqtt_client/src/lib.rs:701]: You tests become much simpler now:

```rust
assert_eq!(trim_null_terminated(vec![155, 156, 157, 158, 0]), vec![155, 156, 157]));
```

File: common/mqtt_client/src/lib.rs [hunk 3]
```diff
@@ -686,4 +693,55 @@ mod tests {
...
+            vec![155, 156, 157, 158]
+        );
+
+        let payload2 = "ab\u{0}";
+        assert_eq!(
+            Message::trim_if_null_terminated(&mut payload2.as_bytes().to_vec()).to_vec(),
```
Comment (mneumann) [inline common/mqtt_client/src/lib.rs:707]: no as_bytes().to_vec()^2 as_bytes() to_vec() orgy needed :D :D :D

**Review by mneumann — COMMENTED** [review]
Your function is unneccessarily creating a copy of the message, regardless of whether it contains a NUL or not! Also, the signature is rather complicated, also see the tests.  If "using" a function starts to get ugly (to_vec, as_bytes etc...), it's a good indication that maybe we can improve the function signature.

**Review by albinsuresh — APPROVED** [review]

**Pull request merged**